### ISSUE TITLE
impossible to upload files to google storage with v0.12.0

### ISSUE THREAD

**Issue by aschmid** [GoogleCloudPlatform/gcloud-node#470]
while 0.11.0 works completely fine i am experiencing huge issues uploading files with the latest (0.12.0) version.

a simple file upload takes forever and might even miss a response.
```
fs.createReadStream('/local/file.txt').pipe(bucket.file('file.txt').createWriteStream());
```

not sure what changed between the versions but it broke something.

**Comment by ryanseys** [comment]
Sorry to hear that! Could you try against the latest code in master. We haven't pushed a new version to npm in a while, so you might have better luck with the most up to date code. Let me know how it goes!

**Comment by aschmid** [comment]
master or 0.12.0 behave the same. 
in my package.json i tried:
```
"gcloud": "0.12.0"
```
and 
```
"gcloud": "*"
```
and both have the same issue.

**Comment by stephenplusplus** [comment]
The `*` will just pull down the latest tag version which is 0.12.0. To get master, do `npm install --save googlecloudplatform/gcloud-node`.

**Comment by aschmid** [comment]
ok i will try

**Comment by aschmid** [comment]
ok so tried again with the latest master branch and it still has the same issue. 
tomorrow i might write a package to prove the case.

**Comment by stephenplusplus** [comment]
It would be great to see a small sample of code. I'm not sure how this would work in 0.11 and not now, but I'm determined to find the cause!

**Comment by ryanseys** [comment]
What version of node are you running? The stack trace seems to hint at the fact that the error occurs outside the realms of our library.

**Comment by aschmid** [comment]
im using node v0.10.38

**Comment by aschmid** [comment]
ok here it is: https://github.com/aschmid/gcloud.storage.test.git

just copy your credentials into gcs_keyfile.json and insert a bucket name and projectid in the config.js object.
after an npm install start with ```./bin/start```

this version doesn't work for me, if i change the gcloud version to 0.11.0 in package.json it works as expected.

**Comment by stephenplusplus** [comment]
After a quick glance, it looks like you're using the end event. Try hooking
on to the complete event.

**Comment by aschmid** [comment]
i tried the complete too but still same result

**Comment by aschmid** [comment]
btw the complete doesnt even work with 0.11.0
thats why i used the 'end'

**Comment by stephenplusplus** [comment]
No problem, I'll take a dive in to the code soon. Thanks for putting the
repo together!

**Comment by stephenplusplus** [comment]
I changed `end` to `complete` and it uploaded the files successfully for me. I'm not sure how we can help, since it's working for us and in our system tests that are run on every change we make to the library.

**Comment by ryanseys** [comment]
0.12.0 is a thing of the past too :) we just recently launched 0.13.1. That being said, if you still have issues I'd suggest a clean install of node and ensure you're not running behind any proxy server or restrictive firewall that might get in the way of uploading!

**Comment by aschmid** [comment]
@ryanseys i had to reset my whole computer 2 days ago so my node environment is as clean as it could be :)
to be honeyst i tried 0.13.1 also and i still experience the same issue... it has to be something else.

**Comment by ryanseys** [comment]
By the looks of the error it's happening at the TLS level so maybe it can't properly negotiate a secure connection. That being said I have no idea why it would work on 0.11 and not 0.13. Funky bug.

**Comment by aschmid** [comment]
yes it is... in any case i was able to work around it but still very annoying.

**Comment by ryanseys** [comment]
Do you have a patch or is your workaround to just use 0.11?

**Comment by aschmid** [comment]
no don't have a patch yet. didn't have time to dig into the actual issue so i used 0.11 for the stuff i needed but obviously i would prefer to be up to date.

**Comment by aschmid** [comment]
would be interesting to know if it is just me or if other ppl have the same issue

**Comment by aschmid** [comment]
i just noticed ```bucket.upload``` works just fine and works through the 20 files within a few seconds.
the example code i supplied takes around 4-5 minutes ending up in an ```end``` event but no files are actually uploaded in the bucket?!
this is very interesting...

**Comment by stephenplusplus** [comment]
bucket.upload is a wrapper around createWriteStream, with some defaults set: https://github.com/GoogleCloudPlatform/gcloud-node/blob/master/lib/storage/bucket.js#L776

Maybe try to mimic that set up as close as you can.

**Comment by stephenplusplus** [comment]
@aschmid going to close as it doesn't seem to be an issue with the library, but something on your end. Feel free to re-open if you can replicate on another machine or provide a test case we can reproduce.

**Comment by aschmid** [comment]
ok good for now. did anybody of you guys try my code? does it work for you?
im strill struggling with this...

**Comment by aschmid** [comment]
ok guys i found the actual issue.

when creating a write stream for the file its absolutely necessary to pass in ```{resumable: <boolean>}``` like so:
```
bucket.file('myfile.jpg').pipe(newFile.createWriteStream({resumable: false}))
```
if this option is not passed in the ```complete``` event never gets called.
i will see if i can submit a working patch in the next few days

**Comment by ryanseys** [comment]
Any patch? :)

**Comment by aschmid** [comment]
no not yet. 
will try to figure out the best way to do this and then supply a patch

**Comment by realguess** [comment]
I had the similar problem with uploading files larger than 5 MB, and the Node process consumes all CPU resource. As suggested by @aschmid, by setting `false` to the `resumable` option, sort of fixing the problem for now:

```js
bucket.upload(src, {
  destination: dest,
  resumable: false
}, callback);
```

**Comment by stephenplusplus** [comment]
Could you be simultaneously uploading files to the same endpoint? I ran parallel file uploads in parallel Node processes with multiple files with the same name, and I would randomly hit an error (code: EPIPE) when two identically named files were being uploaded simultaneously. EPIPE means the socket was closed on the receiving end, which could be explained by the Cloud servers closing the second upload request since another upload was in progress.

**Comment by aschmid** [comment]
@stephenplusplus i am def not uploading files with the same name...

**Comment by stephenplusplus** [comment]
Just to confirm, are these things correct?

1. Using the latest gcloud (0.13.2)
2. Uploading one file this way works: `bucket.upload(path, { resumable: false }, console.log);`
3. Uploading one file this way doesn't work: `bucket.upload(path, console.log);`

**Comment by aschmid** [comment]
@stephenplusplus 
yes i am using the latest gcloud module
i am not using the bucket.upload method because i do not have a local file to pass to the function.
i only have a readstream... so what i do is:

this works:
```
bucketfile = bucket.file(filename).createWriteStream({
          resumable: false,
          validation: false,
          metadata: {
            contentType: mime.lookup(filename)
          }
      })
filestream.pipe(bucketfile)
bucketfile.on 'complete', ->
  console.log 'complete....'
```

while this fails because complete never gets called

```
bucketfile = bucket.file(filename).createWriteStream({
          metadata: {
            contentType: mime.lookup(filename)
          }
      })
filestream.pipe(bucketfile)
bucketfile.on 'complete', ->
  console.log 'complete....'
```

**Comment by stephenplusplus** [comment]
And if you attach an error handler, you still get this: https://github.com/GoogleCloudPlatform/gcloud-node/issues/470#issuecomment-87877603?

**Comment by aschmid** [comment]
i believe so but i would have to try it again...

**Comment by SpainTrain** [comment]
:+1: 

On `gcloud@0.14.0` setting `resumable: false` in options for `createWriteStream` seems to have cleared up the same socket hang up issue for me as well.

**Comment by stephenplusplus** [comment]
I think this could have been from an issue that we [*just* caught](https://github.com/GoogleCloudPlatform/gcloud-node/commit/2478a699a2d7ebb0650093d933781ede2493a1f1). We were storing entire files in a hidden gcloud-node config file, when we only wanted the first 16kb. It would be great if anyone originally experiencing this can try again with 0.16.0.

**Comment by kudos** [comment]
@stephenplusplus I'm still seeing it on 0.16.0, but really happy to have discovered how to work around it!

**Comment by callmehiphop** [comment]
We think this will be resolved in 0.17.0 which is scheduled to be released tomorrow. If anyone experiencing this issue could try to reproduce it, that would be awesome.

**Comment by stephenplusplus** [comment]
This issue is pretty stale; we're 9 releases ahead of when this was reported now. If anyone can reproduce consistently with a test case, I'd love to dig into it more. Until then, `{resumable: false}` seems to be the most reliable fix.
